File: Day_11/main.py
#!/usr/bin/env python3

import numpy as np
from functools import partial

def parse_input(filename):
    with open(filename) as f:
        lines = [list(line.strip()) for line in f.readlines()]
    return np.array(lines)

def expand_universe(universe):
    new_rows = []
    for irow, row in enumerate(universe):
        new_rows.append(row)
        if set(row) == set(["."]):
            new_rows.append(row)

    universe = np.array(new_rows)

    new_cols = []
    for icol, col in enumerate(universe.T):
        new_cols.append(col)
        if set(col) == set(["."]):
            new_cols.append(col)

    return np.array(new_cols).T

def distance_between_galaxies(index1, index2):
    return abs(index2[1] - index1[1]) + abs(index2[0] - index1[0])

def coorindates_of_galaxies(universe):
    coords = []
    for irow, row in enumerate(universe):
        for icol, col in enumerate(row):
            if col == "#":
                coords.append((irow, icol))
    return coords

def answer_part1(filename):
    universe = parse_input(filename)
    universe = expand_universe(universe)
    coords = coorindates_of_galaxies(universe)
    sum_distances = 0
    for icoord, coord in enumerate(coords):
        for j in range(icoord + 1, len(coords)):
            sum_distances += distance_between_galaxies(coord, coords[j])
    return sum_distances


expanded_test = """\
....#........
.........#...
#............
.............
.............
........#....
.#...........
............#
.............
.............
.........#...
#....#......."""

# assert (np.array([list(row) for row in expanded_test.split()]) == expand_universe(parse_input("test_input.txt"))).all()
# assert answer_part1("test_input.txt") == 374
# assert answer_part1("real_input.txt") == 9556896

def expansion_rows_and_columns(universe):
    rows = []
    cols = []
    for irow, row in enumerate(universe):
        if set(row) == set(["."]):
            rows.append(irow)
    for icol, col in enumerate(universe.T):
        if set(col) == set(["."]):
            cols.append(icol)
    return rows, cols

def distance_between_galaxies_with_expansion_info(index1, index2, exanded_rows_and_cols, expansion_factor=1_000_000):
    """It is assumed that exanded_rows_and_cols will be sorted"""
    def filter_function(lower_index, upper_index, index):
        return lower_index <= index <= upper_index or upper_index <= index <= lower_index

    bound_filter = partial(filter_function, index1[0], index2[0])
    filtered_rows = list(filter(bound_filter, exanded_rows_and_cols[0]))
    # print(filtered_rows)
    num_expanded_rows = len(filtered_rows)

    bound_filter = partial(filter_function, index1[1], index2[1])
    filtered_cols = list(filter(bound_filter, exanded_rows_and_cols[1]))
    # print(filtered_cols)
    num_expanded_cols = len(filtered_cols)

    ret = distance_between_galaxies(index1, index2) + (num_expanded_rows + num_expanded_cols) * (expansion_factor - 1)
    # print(f"Distance between {index1} and {index2} is {ret}")
    return ret

assert distance_between_galaxies_with_expansion_info((0, 0), (0, 2), ([], [1]), expansion_factor=1) == 2
assert distance_between_galaxies_with_expansion_info((0, 0), (0, 2), ([], [1]), expansion_factor=10) == 11
assert distance_between_galaxies_with_expansion_info((0, 0), (0, 2), ([], [1]), expansion_factor=100) == 101

assert distance_between_galaxies_with_expansion_info((0, 0), (2, 0), ([1], []), expansion_factor=1) == 2
assert distance_between_galaxies_with_expansion_info((0, 0), (2, 0), ([1], []), expansion_factor=10) == 11
assert distance_between_galaxies_with_expansion_info((0, 0), (2, 0), ([1], []), expansion_factor=100) == 101

assert distance_between_galaxies_with_expansion_info((0, 0), (2, 2), ([1], [1]), expansion_factor=1) == 4
assert distance_between_galaxies_with_expansion_info((0, 0), (2, 2), ([1], [1]), expansion_factor=10) == 22

assert distance_between_galaxies_with_expansion_info((0, 0), (2, 2), ([], [1]), expansion_factor=1) == 4
assert distance_between_galaxies_with_expansion_info((0, 0), (2, 2), ([], [1]), expansion_factor=10) == 13

assert distance_between_galaxies_with_expansion_info((0, 0), (2, 2), ([3], [3]), expansion_factor=100) == 4
assert distance_between_galaxies_with_expansion_info((3, 4), (10, 17), ([5, 9, 10], [11, 12, 13, 17]), expansion_factor=100) == 4 + 9 + 100 * 7

assert distance_between_galaxies_with_expansion_info((0, 3), (1, 7), ([], [2,5,8]), expansion_factor=10) == 14

def answer_part2(filename, expansion_factor):
    universe = parse_input(filename)
    coords = coorindates_of_galaxies(universe)
    exanded_rows_and_cols = expansion_rows_and_columns(universe)
    sum_distances = 0
    for icoord, coord in enumerate(coords):
        for j in range(icoord + 1, len(coords)):
            sum_distances += distance_between_galaxies_with_expansion_info(coord, coords[j], exanded_rows_and_cols, expansion_factor=expansion_factor)
    # print(f"Sum distances: {sum_distances}")
    return sum_distances

assert answer_part2("test_input.txt", expansion_factor=10) == 1030
assert answer_part2("test_input.txt", expansion_factor=100) == 8410
print(answer_part2("real_input.txt", expansion_factor=1_000_000) )

File: Day_11/main.pl
#!/usr/bin/env perl

use strict; # Ensure variables are defined and other nice things
use warnings; # Better diagnostics

sub assert_equal {
    my ($actual, $expected) = @_;
    if ($actual != $expected) {
       die "Assertion failed: $actual != $expected";
    }
}

sub dump_array {
    #
    # Print a 1 or 2-D array to stdout
    #
    my @array = @_;
    foreach my $var (@array) {
        if (ref($var) eq 'ARRAY') {
            print(join('', @$var), "\n");
        }
    }
}

sub parse_input {
    # The below can be done with the following one liner:
    # my @lines = map { [ split '', $_ ] } map { chomp; $_ } <HANDLE>;
    # However, I wanted to break things up to understand them better.

    my ($filename) = @_;
    open(HANDLE, '<', $filename) or die "Could not open file: $!";
    my @lines = <HANDLE>;
    close HANDLE;

    # Remove new lines. Note that this mutates @lines
    map { chomp; $_ } @lines;

    # These two lines are equivalent. The square bracket notation is a more concise way to return
    # a reference to the returned array.
    # @lines = map { my @sp = split('', $_); \@sp } @lines;
    # Split lines into arrays
    @lines = map { [ split('', $_) ] } @lines;
    return \@lines;
}

sub transpose {
    my ($matrix) = @_;
    my @tr;
    my $row_length = @$matrix;
    for my $irow (0..$row_length-1) {
        my $col_length = @{@$matrix[$irow]};
        for my $col (0..$col_length-1) {
            $tr[$col][$irow] = @{@$matrix[$irow]}[$col];
        }
    }
    return \@tr;
}

sub expand_universe {
    my ($universe_ref) = @_;
    my @new_rows;
    # We have to dereference universe_ref first. But then each $row will be a reference to another
    # array
    foreach my $row (@$universe_ref) {
        push @new_rows, $row;
        # TODO: ???
        push @new_rows, $row if scalar(grep { $_ eq '.' } @$row) == scalar(@$row);
    }

    # my @transposed = map { [ map { $_->[$_] } @new_rows ] } 0..$#{$new_rows[0]};
    my $transposed = transpose(\@new_rows);
    my @new_cols;
    foreach my $col (@$transposed) {
        push @new_cols, $col;
        push @new_cols, $col if scalar(grep { $_ eq '.' } @$col) == scalar(@$col);
    }

    # my @result = map { [ map { $_->[$_] } @new_cols ] } 0..$#new_cols;
    return transpose(\@new_cols);
}

sub distance_between_galaxies {
    my ($index1, $index2) = @_;
    return abs($index2->[1] - $index1->[1]) + abs($index2->[0] - $index1->[0]);
}

sub coordinates_of_galaxies {
    my ($universe) = @_;
    my @coords;
    for my $irow (0 .. $#$universe) {
        for my $icol (0 .. $#{$universe->[$irow]}) {
            push @coords, [$irow, $icol] if $universe->[$irow][$icol] eq '#';
        }
    }
    return \@coords;
}

sub answer_part1 {
    my ($filename) = @_;
    # Universe will be a reference to an array of array references (vomit)
    my $universe = parse_input($filename);
    $universe = expand_universe($universe);
    my $coords = coordinates_of_galaxies($universe);
    my $sum_distances = 0;
    for my $icoord (0 .. $#$coords) {
        for my $j ($icoord + 1 .. $#$coords) {
            $sum_distances += distance_between_galaxies($coords->[$icoord], $coords->[$j]);
        }
    }
    return $sum_distances;
}

assert_equal(answer_part1('test_input.txt'), 374);
assert_equal(answer_part1('real_input.txt'), 9556896);

# N.B. I didn't convert part 2 to Perl. I'd had enough by this point!
